### ISSUE TITLE
[Flang][OpenMP] Incorrect DEFAULT(NONE) check for Cray pointers in nested OpenMP regions

### ISSUE THREAD

**Issue by ravurvi20** [llvm/llvm-project#190750]
The following testcase fails during semantic checking when using a Cray pointer inside a nested OpenMP region.
Testcase - 

```fortran
program test
  implicit none
  integer :: i
  real :: X, B
  pointer (P, B)
  
  X = 1.0
  P = LOC(X)
  
  !$omp parallel default(none) shared(P, X)
    !$omp critical
      B = B + 1.0
    !$omp end critical
  !$omp end parallel
  
  print *, "B=", B
end program
```

The testcase raises an error even though the Cray pointer P is declared in the shared clause of the enclosing parallel directive.
Error - 
```
error: The DEFAULT(NONE) clause requires that the Cray Pointer 'p' must be listed in a data-sharing attribute clause
  !$omp critical
        ^^^^^^^^
```
During semantic analysis, when B (the Cray pointee) is referenced inside the critical region:

1. The checker detects that B is a Cray pointee.
2. It retrieves the associated pointer P.
3. It checks whether P has a data-sharing attribute (DSA).

Because the nested critical directive creates a new context, the lookup does not consider the enclosing parallel context where P is declared as shared. This leads to the incorrect diagnostic for the Cray pointee inside the nested region.

**Comment by ravurvi20** [comment]
Changes Merged. Commit Id - [3e7c207](https://github.com/llvm/llvm-project/commit/3e7c207ebe359a7e02b7247b0f24578b17d18eae)
